module3/lessons/many-to-many.md

```diff
@@ -0,0 +1,608 @@
+    expect(playlists[:data][2][:attributes][:name]).to eq("Classic Rock")
+  end
+end
+```
+
+In the setup portion of the test, we are creating the relationships in a couple different ways. Normally, you would want to be more consistent with your syntax, but in this case we want to show a couple different ways to create relationships. Take a minute to review all these different strategies.
+
+Run this test and you should get an error for a missing route. Add it to our routes.
+
+**config/routes.rb**
+
+```ruby
+Rails.application.routes.draw do
+  # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
+
+  # Defines the root path route ("/")
+  # root "articles#indexget "/songs", to: "songs#index"
+namespace :api do
+    namespace :v1 do
+      resources :songs, except: [:new, :edit]
+      resources :artists, only: [:index, :show] do
+        resources :songs, only: [:index, :create], controller: "artist_songs"
+      end
+      resources :playlists, only: [:index]
+      resources :images, only: :index, action: "show"
+    end
+  end
+end
+```
+
+Run the test again and you'll see an error because we don't have a playlists controller, so we'll add the controller with the index action.
+
+**app/controllers/api/v1/playlists_controller.rb**
+
+```ruby
+class Api::V1::PlaylistsController < ApplicationController
+    def index
+        render json: Playlist.all
+    end
+end
+```
+
+Now, we should hit the `pry` in our test, and we can finally see what this data looks like!  One thing that you'll notice is that we get ALL the data about a playlist, including some data we may not care about.  Let's clean that up with a serializer:
+
+**app/serializers/playlist_serializer.rb**
+
+```ruby
+class PlaylistSerializer
+    include JSONAPI::Serializer
+    attributes :name
+end
+```
+
+**app/controllers/api/v1/playlists_controller.rb**
+
+```ruby
+class Api::V1::PlaylistsController < ApplicationController
+    def index
+        render json: PlaylistSerializer.new(Playlist.all)
+    end
+end
+```
+
+Now, when we take a look at the json result in our test, you might be thinking **"wait, what about songs?"**.  With our serializer configured as is, we don't see the representation of the relationship between songs and playlists.  Fortunately, we can fix that with some additional functionality from [JSONAPI::Serializer](https://github.com/jsonapi-serializer/jsonapi-serializer).
+
+```ruby
+class PlaylistSerializer
+    include JSONAPI::Serializer
+    attributes :name
+
+    has_many :songs
+end
+```
+
+**Note**: this does require that a serializer is created for songs!
+
+We also have the ability to add our own custom attributes. What if we wanted an attribute that told us how many songs each playlist had?
+
+```ruby
+class PlaylistSerializer
+    include JSONAPI::Serializer
+    attributes :name
+
+    has_many :songs
+
+    attribute :num_songs do |object|
+        object.songs.count
```

Comment:
I think keeping this section in the lesson plan makes sense, because this serializer functionality is not new to them, just the idea of adding relationships to it. 

In the past, though, students always point out that this is a MVC violation. We don't want to use an AR method like `.count` in the presentation logic (this is something we drilled in so much in BE mod 2). So it'd be great to update this and have it calling a Playlist model method called `count_songs` or something like that.

This did always spark some good conversation though, as devs violate this MVC rule all the time and I've seen `.count` and other simple query methods all over views... so IDK!

Reply:
Totally agree on the songs.count being an MVC infraction.  I like that this has come up in conversation - I'll add a discussion question on this!  I also included a challenge at the end that asks them to recreate similar functionality without using enumerables in the serializer.